.github/workflows/build.yml

```diff
@@ -78,13 +78,39 @@ jobs:
       - name: Gradle build
         uses: gradle/gradle-build-action@v2
         with:
-          arguments: --refresh-dependencies --stacktrace --scan clean build
+          arguments: --refresh-dependencies --stacktrace --scan clean build -x spotlessCheck
       - name: SonarCloud analysis
         if: ${{ matrix.os == 'ubuntu' }}
         env:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           SONAR_TOKEN: "61ab2579215aa8a0024a2f9368fc1298fdecfd18"
         run: ./gradlew jacocoTestReport sonarqube --stacktrace -i
+      - name: Test report
+        uses: mikepenz/action-junit-report@v2
+        if: always() # always run even if the previous step fails
+        with:
+          check_name: "Test report for Java ${{ matrix.java }} on ${{ matrix.os }}"
+          report_paths: '**/build/test-results/test/TEST-*.xml'
+
+  # Code format check
+  code-format-check:
+    runs-on: ubuntu-latest
+    needs: validation
```

Comment:
Maybe we now want to run this cheap `code-format-check` before `basic`, i.e. let `basic` need this?

Reply:
The idea was to run all on parallel, so the feedback for the contributor is faster.

So it can be done in one swoop, and not two build iterations.

But what I thought is, maybe run spotless and one is, instead of 3 - and continue after those 2 build steps. Would this be a good idea?

The issue I see, is that it would make the waiting time for contributors longer, to get a full feedback on their contribution from CI system

Reply:
But is the feedback faster than before? Currently we have:

```text
validation --> basic (incl. Spotless) --> ...
```

With this PR:

```text
validation --> basic (excl. Spotless) --> ...
            |
            -> code-format-check
```

Before, `basic` would fail (fast) in case of a Spotless error. Now, `basic` passes but `code-format-check` would fail, probably just as fast as `basic` before?

The motivation behind my suggestion was to save computation from our (I believe) 2,000 minutes/month limit. That is, don't run the `basic` matrix until spotless errors have been resolved.

Reply:
But from a contributor perspective there are two different paths:

Approach with separate spotless:
1. Open pr
2. Builds fails on spotless
3. Fix code
4. Wait for build
5. Build might fail for unit tests
6. Fix unit tests

Versus
In parallel:
1. Open pr
2. Build fails on spotless
3. Build fails on tests
4. Fix both

If we want to trim it, I would suggest to have just one basic job just for one environment in parallel with spotless. Most likely the tests will not fail on other OS (besides some special extension) and run them afterwards.

I think this would unify both worlds, WDYT? Less computation time, and fast feedback for all basic needs.

Reply:
I see your point, it's a trade-off.

When you say "just one basic job", you mean something like?

```
validation --> basic (no matrix, ubuntu only) --> full-featured (merge-ready label)
            |
            -> code-format-check
```

?

Reply:
sorry i was only on my mobile so i had issue properly formatting. but yes something like you posted or:

```
validation --> basic (no matrix, ubuntu only) -
            |                                                          |-> matrix build (3os) --> full-featured (merge-ready label)
            -> code-format-check                      -
```

so 3 phase:
1. Security (gradle wrapper validation)
2. Basic Verification and checks (currently spotless and basic build, but could be increases)
3. Extended Verification
4. Fully Fledged ultimative pre merge stage

Reply:
How would `basic` and the Ubuntu-part of `matrix build` differ? Would it be sufficient to run the matrix only with macOS and Windows, if `basic` already covers Ubuntu? Example:

```mermaid
flowchart LR
    v[validation]
    b[basic]
    c[code-format-check]
    x[x-os]
    f[full-featured]

    v --> b
    v --> c
    b --> x
    c --> x
    x --> f
```

(Just rembered that GitHub now supports [Mermaid](https://mermaid-js.github.io/).)

Reply:
i do think that this might be the way to go:

- we gain fast feedback as a contributor (to be honest there are seldomly issues with other OS's)
- we reduce the build times, we just have 3 jobs which will do all kind of verifications - and the rest at a later stage
- we still have full verification at the end but fast and bundled feedback

Reply:
I will mark this as resolved, since there is now #614 for further discussion.